src/coreclr/System.Private.CoreLib/src/System/Reflection/MethodBase.CoreCLR.cs

```diff
@@ -62,30 +65,59 @@ internal virtual Type[] GetParameterTypes()
+                "MaxStackAllocArgCount not properly defined.");
 
-            ParameterInfo[]? p = null;
-            for (int i = 0; i < parameters.Length; i++)
+            Span<object?> copyOfParameters = default;
+
+            if (parameters is not null)
             {
-                object arg = parameters[i];
-                RuntimeType argRT = sig.Arguments[i];
+                // copy the arguments into a temporary buffer (or a new array) so we detach from any user changes
+                copyOfParameters = (parameters.Length <= StackAllocedArguments.MaxStackAllocArgCount)
+                        ? MemoryMarshal.CreateSpan(ref stackArgs._arg0, parameters.Length)
+                        : new Span<object?>(new object?[parameters.Length]);
 
-                if (arg == Type.Missing)
+                ParameterInfo[]? p = null;
+                for (int i = 0; i < parameters.Length; i++)
                 {
-                    p ??= GetParametersNoCopy();
-                    if (p[i].DefaultValue == System.DBNull.Value)
-                        throw new ArgumentException(SR.Arg_VarMissNull, nameof(parameters));
-                    arg = p[i].DefaultValue!;
+                    object? arg = parameters[i];
+                    RuntimeType argRT = sig.Arguments[i];
+
+                    if (arg == Type.Missing)
+                    {
+                        p ??= GetParametersNoCopy();
+                        if (p[i].DefaultValue == System.DBNull.Value)
+                            throw new ArgumentException(SR.Arg_VarMissNull, nameof(parameters));
+                        arg = p[i].DefaultValue!;
+                    }
+                    copyOfParameters[i] = argRT.CheckValue(arg, binder, culture, invokeAttr);
                 }
-                copyOfParameters[i] = argRT.CheckValue(arg, binder, culture, invokeAttr)!;
             }
 
             return copyOfParameters;
         }
+
+        // Helper struct to avoid intermediate object[] allocation in calls to the native reflection stack.
+        // Typical usage is to define a local of type default(StackAllocedArguments), then pass 'ref theLocal'
+        // as the first parameter to CheckArguments. CheckArguments will try to utilize storage within this
+        // struct instance if there's sufficient space; otherwise CheckArguments will allocate a temp array.
+        private protected struct StackAllocedArguments
+        {
+            internal const int MaxStackAllocArgCount = 8;
+            internal object? _arg0;
+#pragma warning disable CA1823, CS0169, IDE0051 // accessed via 'CheckArguments' ref arithmetic
+            private object? _arg1;
+            private object? _arg2;
+            private object? _arg3;
+            private object? _arg4;
+            private object? _arg5;
+            private object? _arg6;
+            private object? _arg7;
+#pragma warning restore CA1823, CS0169, IDE0051
+        }
```

Comment:
We now have a few of these floating around, e.g.
https://github.com/dotnet/runtime/blob/79ae74f5ca5c8a6fe3a48935e85bd7374959c570/src/libraries/Common/src/Interop/Windows/SspiCli/SSPIWrapper.cs#L218-L224
https://github.com/dotnet/runtime/blob/79ae74f5ca5c8a6fe3a48935e85bd7374959c570/src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexReplacement.cs#L97-L103
and other places we'd wanted to but just didn't go to the lengths of creating these (there's also the additional zero'ing overhead that comes if you overestimate). At some point, we should think about doing something more general, whether it be enabling runtime support for stackalloc to work with reference types, or language support to create these types for stackallocs with const sizes, or a set of helper types we simply share either in Common or even publicly.